### ISSUE TITLE
Array#product is quite confusing.

### ISSUE THREAD

**Issue by yxhuvud** [crystal-lang/crystal#3790]
Enumerable#product and #product(initial : Numeric) each will try to multiply the contents of whatever enumerable it is. There is also Array#product(ary : Array(U)) which will produce the set theoretic product of the two arrays. 

This is confusing for several reasons:
1: product does totally different things depending on input.
2: There is no way to do a product between actual Sets without converting them to an array. 

I'd suggest that at the very least one the second of these issues need to go away. Possibly by moving Array#product into Enumerable (perhaps renamed).

Perhaps it would also be nice to make product (and zip etc) into taking multiple arguments - being able to get flat structures like s1 xs2 x s3 without having to flatten any step would be very convenient in some cases.

**Comment by asterite** [comment]
What is a product between sets?

**Comment by radiospiel** [comment]
Should this even be a Array instance method? Couldn't a `Math.cartesian_product(ary, other_ary)` serve the same purpose without polluting Array's namespace?

**Comment by iSarCasm** [comment]
Any updates? @asterite

**Comment by HertzDevil** [comment]
#3794 wouldn't even work if any of the arguments is infinite, like `Range`.

Would the size issue be fixed by moving `#product` to `Indexable` instead of `Enumerable`?

**Comment by asterite** [comment]
I personally don't think this is confusing. A product between two collections is one thing, then the product of the elements of a collection, with an initial value, is another thing. That's the beauty of being able to overload things. The term "product" is overloaded, and so the definitions in the language.

I would try to move product to Indexable, and then add product to set, but that's it. I wouldn't use a different name.

**Comment by straight-shoota** [comment]
@asterite I strongly disagree. Overloads are useful for having slightly different behaviour depending on arguments.
But it's a really bad idea to overload the same method name with two completely different operations. That's totally unacceptable.

**Comment by straight-shoota** [comment]
Number 1. has been resolved with #10013. Number 2. (product of sets is still open).

**Comment by straight-shoota** [comment]
We could surely add an overload for `each_cartesian` that accepts `Set`s explicitly. That would at least offer a solution for pure `Set` products. Mixing with other collection types would be difficult though.

Alternatively we could consider the introduction of another collection type that indicates an `Enumerable` with stable elements but which can't be accessed by index.
*Stable* can probably be translated to being backed by a concrete collection of elements, which probably implies mutability. So this could be the equivalent of `Indexable::Mutable` for `Enumerable` (maybe `Enumerable::Mutable`? https://github.com/crystal-lang/crystal/issues/5142#issuecomment-873384617). It's also a resemblance of Java's [`Collection`](https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html) type.
Tangentially relevant is also the problem with `Enumerable#size` (#10014, #10075) which affects the same distinction.

**Comment by HertzDevil** [comment]
It is perfectly possible to e.g. implement a set as a persistent data structure as other FP languages do, so stability does not imply mutability. Likewise stability does not imply a finite size (e.g. `1..` is stable but not finite).

We want stability and finiteness. A minimum definition would then be:

```crystal
module Collection(T)
  include Enumerable(T)

  # moved from `Indexable`
  abstract def size : Int32
end

module Indexable(T)
  include Collection(T)
end

struct Set(T)
  include Collection(T)
end

class Hash(K, V)
  include Collection({K, V})
end
```

Note that `Collection`'s stability guarantee is purely semantic. With this `cartesian_product` may accept `Collection` arguments, and we may assert that `Collection#to_a` and `#size` never invalidate the receiver.
